Mount Devices

Trying to mount an arduino to a container runned by dobi Fails

```
mount=device:
    bind: /dev/ttyACM0
    path: /dev/ttyACM0

job=watch:
    use: watch-builder
    interactive: true
    net-mode: host
    ports: ["5000:5000"]
    mounts: [source,machine-id,device]
    env:
      - "POWER_JWT={env.POWER_JWT:}"
    description: "Watch for code changes and run the unit tests"
```


The same container created by dobi, run with plain docker using the --device flag:

```
cesco@cesco-desktop: ~/code/go/src/github.com/cescoferraro/power on master [+!?]
$ docker run -p 5000:5000 -v /var/lib/dbus/machine-id:/var/lib/dbus/machine-id  -v $PWD:/go/src/github.com/cescoferraro/power --device /dev/ttyACM0 --net=host  cescoferraro/power:watch 
[00] Starting service
[00] 2016/10/22 00:38:52 Failed to read config Unsupported Config Type ""
[00] 2016/10/22 00:38:52 verbose enabled
[00] 2016/10/22 00:38:52 ***********VIPER*************
[00] 2016/10/22 00:38:52 ping: true
[00] 2016/10/22 00:38:52 channels: 8
[00] 2016/10/22 00:38:52 owner: francescoaferraro@gmail.com
[00] 2016/10/22 00:38:52 api: https://api.cescoferraro.xyz/iot/devices
[00] 2016/10/22 00:38:52 dev-api: http://localhost:9000/iot/devices
[00] 2016/10/22 00:38:52 serial_port: /dev/ttyACM0
```

I am trying to set a single device on job/run.go and it works like this. I was wondering how should we implement this. ["host:container:cgroup"] or  ["same:cgroup"] or ["same"] with full permissions?

```
job=binary:
    use: builder
    artifact: ./dist/bin/
    devices: "/dev/ttyACM0"


	device := docker.Device{
		PathInContainer:t.config.Devices,
		PathOnHost:t.config.Devices,
		CgroupPermissions:"rwm"}

	portBinds, exposedPorts := asPortBindings(t.config.Ports)
	// TODO: only set Tty if running in a tty
	opts := docker.CreateContainerOptions{
		Name: name,
		Config: &docker.Config{
			Cmd:          t.config.Command.Value(),
			Image:        imageName,
			User:         t.config.User,
			OpenStdin:    interactive,
			Tty:          interactive,
			AttachStdin:  interactive,
			StdinOnce:    interactive,
			AttachStderr: true,
			AttachStdout: true,
			Env:          t.config.Env,
			Entrypoint:   t.config.Entrypoint.Value(),
			WorkingDir:   t.config.WorkingDir,
			ExposedPorts: exposedPorts,
		},
		HostConfig: &docker.HostConfig{
			Binds:        t.bindMounts(ctx),
			Privileged:   t.config.Privileged,
			NetworkMode:  t.config.NetMode,
			Devices:  []docker.Device{device},
			PortBindings: portBinds,
		},
	}

```

Funny thing its the I got it to work before...
```
cesco@cesco-desktop: ~/code/go/src/github.com/dnephin/dobi on feature/devices [!]
$ go run main.go binary
[ERROR] Failed to load config from "dobi.yaml": Invalid config for resource "job=binary":
Error at job=binary.devices.0: item in the list is of wrong type "map", expected "config.Dev"
exit status 1


job=binary:
    use: builder
    artifact: ./dist/bin/
    devices:
      - host: /dev/ttyACM0
        container: /dev/ttyACM0
        permissions: rwm
    mounts: [source, dist]
    command: script/build


	//Devices  Maps the host devices you want to connect to a container
	Devices []Dev
	// Ports Publish ports to the host
	// type: list of 'host_port:container_port'
	Ports []string
	dependent
	describable
}

type Dev struct {
	Host  string
	Container string
	Permissions string
}


```

--- a/config/job.go
+++ b/config/job.go
@@ -73,7 +73,9 @@ type JobConfig struct {
 	WorkingDir string
 	// User Username or UID to use in the container. Format ``user[:group]``.
 	User string
-	//Devices  Maps the host devices you want to connect to a container
+	// Devices  Maps the host devices you want to connect to a container
+	// type: list of device specs
+	// example: ``{Host: /dev/fb0, Container: /dev/fb0, Permissions: rwm}``
 	Devices []Device
 	// Ports Publish ports to the host
 	// type: list of 'host_port:container_port'
@@ -82,7 +84,7 @@ type JobConfig struct {
 	describable
 }
 
-//Device is the defined structure to attach host devices to containers
+// Device is the defined structure to attach host devices to containers
 type Device struct {
 	Host        string
 	Container   string

--- a/config/job.go
+++ b/config/job.go
@@ -73,13 +73,13 @@ type JobConfig struct {
 	WorkingDir string
 	// User Username or UID to use in the container. Format ``user[:group]``.
 	User string
-	// Devices  Maps the host devices you want to connect to a container
-	// type: list of device specs
-	// example: ``{Host: /dev/fb0, Container: /dev/fb0, Permissions: rwm}``
-	Devices []Device
 	// Ports Publish ports to the host
 	// type: list of 'host_port:container_port'
 	Ports []string
+	// Devices Maps the host devices you want to connect to a container
+	// type: list of device specs
+	// example: ``{Host: /dev/fb0, Container: /dev/fb0, Permissions: rwm}``
+	Devices []Device
 	dependent
 	describable
 }

--- a/config/job.go
+++ b/config/job.go
@@ -73,11 +73,20 @@ type JobConfig struct {
 	WorkingDir string
 	// User Username or UID to use in the container. Format ``user[:group]``.
 	User string
+	//Devices  Maps the host devices you want to connect to a container
+	Devices []Device
 	// Ports Publish ports to the host
 	// type: list of 'host_port:container_port'
 	Ports []string
 	dependent
 	describable
+}
+
+//Device is the defined structure to attach host devices to containers
+type Device struct {
+	Host        string
+	Container   string
+	Permissions string
 }
 
 // Dependencies returns the list of implicit and explicit dependencies

--- a/tasks/job/run.go
+++ b/tasks/job/run.go
@@ -207,15 +207,6 @@ func (t *Task) runContainer(ctx *context.ExecuteContext) error {
 
 func (t *Task) createOptions(ctx *context.ExecuteContext, name string) docker.CreateContainerOptions {
 	interactive := t.config.Interactive
-	var dockerdevices []docker.Device
-	for _, dev := range t.config.Devices {
-		dockerdevices = append(dockerdevices,
-			docker.Device{
-				PathInContainer:   dev.Container,
-				PathOnHost:        dev.Host,
-				CgroupPermissions: dev.Permissions,
-			})
-	}
 
 	imageName := image.GetImageName(ctx, ctx.Resources.Image(t.config.Use))
 	t.logger().Debugf("Image name %q", imageName)
@@ -244,11 +235,30 @@ func (t *Task) createOptions(ctx *context.ExecuteContext, name string) docker.Cr
 			Privileged:   t.config.Privileged,
 			NetworkMode:  t.config.NetMode,
 			PortBindings: portBinds,
-			Devices:      dockerdevices,
+			Devices:      getDevices(t.config.Devices),
 		},
 	}
 	opts = provideDocker(opts)
 	return opts
+}
+
+func getDevices(devices []config.Device) []docker.Device {
+	var dockerdevices []docker.Device
+	for _, dev := range devices {
+		if len(dev.Container) == 0 {
+			dev.Container = dev.Host
+		}
+		if len(dev.Permissions) == 0 {
+			dev.Container = "rwm"
+		}
+		dockerdevices = append(dockerdevices,
+			docker.Device{
+				PathInContainer:   dev.Container,
+				PathOnHost:        dev.Host,
+				CgroupPermissions: dev.Permissions,
+			})
+	}
+	return dockerdevices
 }
 
 func asPortBindings(ports []string) (map[docker.Port][]docker.PortBinding, map[docker.Port]struct{}) {

--- a/tasks/job/run.go
+++ b/tasks/job/run.go
@@ -207,6 +207,15 @@ func (t *Task) runContainer(ctx *context.ExecuteContext) error {
 
 func (t *Task) createOptions(ctx *context.ExecuteContext, name string) docker.CreateContainerOptions {
 	interactive := t.config.Interactive
+	var dockerdevices []docker.Device
+	for _, dev := range t.config.Devices {
+		dockerdevices = append(dockerdevices,
+			docker.Device{
+				PathInContainer:   dev.Container,
+				PathOnHost:        dev.Host,
+				CgroupPermissions: dev.Permissions,
+			})
+	}
 
 	imageName := image.GetImageName(ctx, ctx.Resources.Image(t.config.Use))
 	t.logger().Debugf("Image name %q", imageName)
@@ -235,6 +244,7 @@ func (t *Task) createOptions(ctx *context.ExecuteContext, name string) docker.Cr
 			Privileged:   t.config.Privileged,
 			NetworkMode:  t.config.NetMode,
 			PortBindings: portBinds,
+			Devices:      dockerdevices,
 		},
 	}
 	opts = provideDocker(opts)

--- a/tasks/job/run.go
+++ b/tasks/job/run.go
@@ -245,11 +245,11 @@ func (t *Task) createOptions(ctx *context.ExecuteContext, name string) docker.Cr
 func getDevices(devices []config.Device) []docker.Device {
 	var dockerdevices []docker.Device
 	for _, dev := range devices {
-		if len(dev.Container) == 0 {
+		if dev.Container == "" {
 			dev.Container = dev.Host
 		}
-		if len(dev.Permissions) == 0 {
-			dev.Container = "rwm"
+		if dev.Permissions == "" {
+			dev.Permissions = "rwm"
 		}
 		dockerdevices = append(dockerdevices,
 			docker.Device{

